File: sessions/week-01-two-pointers/algorithms/javascript/reverseLinkedList.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

function reverseLinkedList(head) {
    let prev = null;
    let current = head;
    while (current !== null) {
        let nextNode = current.next;
        current.next = prev;
        prev = current;
        current = nextNode;
    }
    return prev;
}

function buildLinkedList(values) {
    const head = new Node(values[0]);
    let current = head;
    for (let i = 1; i < values.length; i++) {
        current.next = new Node(values[i]);
        current = current.next;
    }
    return head;
}

function printList(head) {
    let current = head;
    while (current) {
        process.stdout.write(current.value + " -> ");
        current = current.next;
    }
    console.log("null");
}

const head = buildLinkedList([1, 2, 3, 4, 5]);
console.log("Original list:");
printList(head);

const reversedHead = reverseLinkedList(head);
console.log("Reversed list:");
printList(reversedHead);

File: sessions/week-01-two-pointers/algorithms/java/ReverseLinkedList.java
class Node {
    int value;
    Node next;
    Node(int value) {
        this.value = value;
    }
}

public class ReverseLinkedList {
    public static Node reverseList(Node head) {
        Node prev = null;
        Node current = head;
        while (current != null) {
            Node nextNode = current.next;
            current.next = prev;
            prev = current;
            current = nextNode;
        }
        return prev;
    }

    public static void printList(Node head) {
        while (head != null) {
            System.out.print(head.value + " -> ");
            head = head.next;
        }
        System.out.println("null");
    }

    public static void main(String[] args) {
        Node head = new Node(1);
        head.next = new Node(2);
        head.next.next = new Node(3);
        head.next.next.next = new Node(4);

        System.out.println("Original list:");
        printList(head);

        Node reversed = reverseList(head);
        System.out.println("Reversed list:");
        printList(reversed);
    }
}
